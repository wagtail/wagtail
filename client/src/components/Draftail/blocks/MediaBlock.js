import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Icon } from 'draftail';

import Tooltip from '../Tooltip/Tooltip';
import Portal from '../../Portal/Portal';

// Constraints the maximum size of the tooltip.
const OPTIONS_MAX_WIDTH = 300;
const OPTIONS_SPACING = 70;
const TOOLTIP_MAX_WIDTH = OPTIONS_MAX_WIDTH + OPTIONS_SPACING;

/**
 * Editor block to preview and edit images.
 */
class MediaBlock extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showTooltipAt: null,
    };

    this.openTooltip = this.openTooltip.bind(this);
    this.closeTooltip = this.closeTooltip.bind(this);
    this.renderTooltip = this.renderTooltip.bind(this);
  }

  openTooltip(e) {
    const trigger = e.target.closest('[data-draftail-trigger]');

    // Click is within the tooltip.
    if (!trigger) {
      return;
    }

    const container = trigger.closest('[data-draftail-editor-wrapper]');
    const containerRect = container.getBoundingClientRect();
    const rect = trigger.getBoundingClientRect();
    const maxWidth = trigger.parentNode.offsetWidth - rect.width;

    this.setState({
      showTooltipAt: {
        container: container,
        top: rect.top - containerRect.top - (document.documentElement.scrollTop || document.body.scrollTop),
        left: rect.left - containerRect.left - (document.documentElement.scrollLeft || document.body.scrollLeft),
        width: rect.width,
        height: rect.height,
        direction: maxWidth >= TOOLTIP_MAX_WIDTH ? 'left' : 'top-left',
      },
    });
  }

  closeTooltip() {
    this.setState({ showTooltipAt: null });
  }

  renderTooltip() {
    const { children } = this.props;
    const { showTooltipAt } = this.state;

    return (
      <Portal
        node={showTooltipAt.container}
        onClose={this.closeTooltip}
        closeOnClick
        closeOnType
        closeOnResize
      >
        <Tooltip target={showTooltipAt} direction={showTooltipAt.direction}>
          <div style={{ maxWidth: OPTIONS_MAX_WIDTH }}>{children}</div>
        </Tooltip>
      </Portal>
    );
  }

  render() {
    const { blockProps, src, alt } = this.props;
    const { showTooltipAt } = this.state;
    const { entityType } = blockProps;

    return (
      <button
        type="button"
        tabIndex={-1}
        className="MediaBlock"
        onClick={this.openTooltip}
        data-draftail-trigger
      >
        <span className="MediaBlock__icon-wrapper" aria-hidden>
          <Icon icon={entityType.icon} className="MediaBlock__icon" />
        </span>

        <img className="MediaBlock__img" src={src} alt={alt} width="256" />

        {showTooltipAt && this.renderTooltip()}
      </button>
    );
  }
}

MediaBlock.propTypes = {
  blockProps: PropTypes.shape({
    entityType: PropTypes.object.isRequired,
  }).isRequired,
  src: PropTypes.string,
  alt: PropTypes.string,
  children: PropTypes.node.isRequired,
};

MediaBlock.defaultProps = {
  src: null,
  alt: '',
};

export default MediaBlock;
