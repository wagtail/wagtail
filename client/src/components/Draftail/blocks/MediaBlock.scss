.MediaBlock {
    display: inline-block;
    position: relative;
    border: 0;
    padding: 0;
    cursor: pointer;

    &__icon-wrapper {
        position: absolute;
        top: 0;
        right: 0;
        background: $draftail-editor-chrome;
        color: $draftail-editor-chrome-text;
        line-height: 1;
        padding: $controls-spacing * 2 $controls-spacing * 3;
        pointer-events: none;

        .icon {
            @include svg-icon();
        }
    }

    &__icon {
        $icon-size: 1.5rem;
        width: $icon-size;
        height: $icon-size;
        font-size: $icon-size;
    }

    @mixin invalid-image-fallback {
        min-width: 256px;
        min-height: 50px;
        object-fit: contain;
        background-color: $color-grey-1;
    }

    &__img {
        @include invalid-image-fallback;
        display: block;
        width: 256px;
        height: auto;
        pointer-events: none;
    }
}
