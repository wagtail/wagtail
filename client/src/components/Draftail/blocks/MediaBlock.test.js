import React from 'react';
import { shallow, mount } from 'enzyme';

import MediaBlock from '../blocks/MediaBlock';

describe('MediaBlock', () => {
  it('renders', () => {
    expect(
      shallow(
        <MediaBlock
          src="example.png"
          alt=""
          blockProps={{
            entityType: {
              icon: '#icon-test',
            },
            entity: {
              getData: () => ({
                src: 'example.png',
              }),
            },
          }}
        >
          Test
        </MediaBlock>
      )
    ).toMatchSnapshot();
  });

  it('no data', () => {
    expect(
      shallow(
        <MediaBlock
          src=""
          alt=""
          blockProps={{
            entityType: {
              icon: '#icon-test',
            },
            entity: {
              getData: () => ({}),
            },
          }}
        >
          Test
        </MediaBlock>
      )
    ).toMatchSnapshot();
  });

  describe('tooltip', () => {
    let target;
    let wrapper;

    beforeEach(() => {
      target = document.createElement('div');
      target.setAttribute('data-draftail-trigger', true);
      document.body.appendChild(target);
      document.body.setAttribute('data-draftail-editor-wrapper', true);

      wrapper = mount(
        <MediaBlock
          src="example.png"
          alt=""
          blockProps={{
            entityType: {
              icon: '#icon-test',
            },
            entity: {
              getData: () => ({
                src: 'example.png',
              }),
            },
          }}
        >
          <div id="test">Test</div>
        </MediaBlock>
      );
    });

    it('opens', () => {
      wrapper.simulate('click', { target });

      expect(
        wrapper
          .find('Portal > Portal').prop('containerInfo')
      ).toMatchSnapshot();
    });

    it('click in tooltip', () => {
      wrapper.simulate('click', { target });

      jest.spyOn(target, 'getBoundingClientRect');

      wrapper.simulate('click', { target: document.querySelector('#test') });

      expect(target.getBoundingClientRect).not.toHaveBeenCalled();
    });

    it('large viewport', () => {
      target.getBoundingClientRect = () => ({
        top: 0,
        left: 0,
        width: -600,
        height: 0,
      });

      wrapper.simulate('click', { target });

      expect(
        wrapper
          .find('.Tooltip')
          .prop('className')
      ).toBe('Tooltip Tooltip--left');
    });

    it('closes', () => {
      jest.spyOn(target, 'getBoundingClientRect');

      expect(wrapper.state('showTooltipAt')).toBe(null);

      wrapper.simulate('click', { target });

      expect(wrapper.state('showTooltipAt')).toMatchObject({
        top: 0,
        left: 0,
      });
      expect(target.getBoundingClientRect).toHaveBeenCalled();

      wrapper.instance().closeTooltip();

      expect(wrapper.state('showTooltipAt')).toBe(null);
    });
  });
});
