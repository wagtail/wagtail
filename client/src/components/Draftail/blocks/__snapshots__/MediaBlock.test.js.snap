// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaBlock no data 1`] = `
<button
  className="MediaBlock"
  data-draftail-trigger={true}
  onClick={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    aria-hidden={true}
    className="MediaBlock__icon-wrapper"
  >
    <Icon
      className="MediaBlock__icon"
      icon="#icon-test"
      title={null}
    />
  </span>
  <img
    alt=""
    className="MediaBlock__img"
    src=""
    width="256"
  />
</button>
`;

exports[`MediaBlock renders 1`] = `
<button
  className="MediaBlock"
  data-draftail-trigger={true}
  onClick={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    aria-hidden={true}
    className="MediaBlock__icon-wrapper"
  >
    <Icon
      className="MediaBlock__icon"
      icon="#icon-test"
      title={null}
    />
  </span>
  <img
    alt=""
    className="MediaBlock__img"
    src="example.png"
    width="256"
  />
</button>
`;

exports[`MediaBlock tooltip opens 1`] = `
<div>
  <div
    class="Tooltip Tooltip--top-left"
    role="tooltip"
    style="top: 0px; left: 0px;"
  >
    <div
      style="max-width: 300px;"
    >
      <div
        id="test"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;
